src/os/kqueue.rs

```diff
@@ -0,0 +1,329 @@
+    /// instead of a socket. See the implementors of the [`Filter`] trait for more information.
+    ///
+    /// # Examples
+    ///
+    /// ```no_run
+    /// use polling::{Poller, PollMode};
+    /// use polling::os::kqueue::{Filter, PollerKqueueExt, Signal};
+    ///
+    /// let poller = Poller::new().unwrap();
+    ///
+    /// // Register the SIGINT signal.
+    /// poller.add_filter(Signal::new(libc::SIGINT), 0, PollMode::Oneshot).unwrap();
+    ///
+    /// // Re-register with a different key.
+    /// poller.modify_filter(Signal::new(libc::SIGINT), 1, PollMode::Oneshot).unwrap();
+    ///
+    /// // Wait for the signal.
+    /// let mut events = vec![];
+    /// poller.wait(&mut events, None).unwrap();
+    /// # let _ = events;
+    /// ```
+    fn modify_filter(&self, filter: F, key: usize, mode: PollMode) -> io::Result<()>;
+
+    /// Remove a filter from the poller.
+    ///
+    /// This is used to remove filters that were previously added with
+    /// [`add_filter`](PollerKqueueExt::add_filter).
+    ///
+    /// # Examples
+    ///
+    /// ```no_run
+    /// use polling::{Poller, PollMode};
+    /// use polling::os::kqueue::{Filter, PollerKqueueExt, Signal};
+    ///
+    /// let poller = Poller::new().unwrap();
+    ///
+    /// // Register the SIGINT signal.
+    /// poller.add_filter(Signal::new(libc::SIGINT), 0, PollMode::Oneshot).unwrap();
+    ///
+    /// // Remove the filter.
+    /// poller.delete_filter(Signal::new(libc::SIGINT)).unwrap();
+    /// ```
+    fn delete_filter(&self, filter: F) -> io::Result<()>;
+}
+
+impl PollerSealed for Poller {}
+
+impl<F: Filter> PollerKqueueExt<F> for Poller {
+    fn add_filter(&self, filter: F, key: usize, mode: PollMode) -> io::Result<()> {
+        // No difference between adding and modifying in kqueue.
+        self.modify_filter(filter, key, mode)
+    }
+
+    fn modify_filter(&self, filter: F, key: usize, mode: PollMode) -> io::Result<()> {
+        // Convert the filter into a kevent.
+        let event = filter.filter(libc::EV_ADD | mode_to_flags(mode), key);
+
+        // Modify the filter.
+        self.poller.submit_changes([event])
+    }
+
+    fn delete_filter(&self, filter: F) -> io::Result<()> {
+        // Convert the filter into a kevent.
+        let event = filter.filter(libc::EV_DELETE, 0);
+
+        // Delete the filter.
+        self.poller.submit_changes([event])
+    }
+}
+
+/// A filter that can be registered into a `kqueue`.
+pub trait Filter: FilterSealed {}
+
+unsafe impl<T: FilterSealed + ?Sized> FilterSealed for &T {
+    fn filter(&self, flags: FilterFlags, key: usize) -> libc::kevent {
+        (**self).filter(flags, key)
+    }
+}
+
+impl<T: Filter + ?Sized> Filter for &T {}
+
+/// Monitor this signal number.
+///
+/// No matter what `PollMode` is specified, this filter will always be
+/// oneshot-only.
+#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct Signal(c_int);
```

Comment:
it would make more sense to just make the single wrapped field `pub` and omit the construct/destruct methods.